tom/packages.py

```diff
@@ -0,0 +1,96 @@
+        for branch in branches.split(','):
+            self.slack.reply("Updating packages mapping for " + branch)
+            result = {'agent':{},'hub':{}}
+            for product in ['community','enterprise']:
+                releases_data = requests.get('https://cfengine.com/release-data/%s/releases.json' % product).json()
+                release_url, version = next((release['URL'], release['version']) for release in releases_data['releases']
+                                            if "lts_branch" in release and release["lts_branch"] == branch
+                                            and "latest_on_branch" in release and release["latest_on_branch"])
+                packages = self.collect_packages(release_url)
+                # sort packages into agent/hub ones
+                if product == 'community':
+                    # for community, they are the same
+                    hub_packages = agent_packages = packages
+                else:
+                    hub_packages = {platform:value for (platform,value) in packages.items() if '_HUB_' in platform}
+                    agent_packages = {platform:value for (platform,value) in packages.items() if '_HUB_' not in platform}
+                result['agent'][product] = agent_packages
+                result['hub'][product] = hub_packages
+            packages_mapping['packages'][version] = result
+            repo.put_file(packages_mapping_file_path, json.dumps(packages_mapping, indent=2))
+            # update version in upgrade_from.json
+            branch_nox=re.sub('\\.x$', '', branch)
+            upgrade_from_file_path = 'deployment_tests/upgrade_from.json'
+            upgrade_from = json.loads(repo.get_file(upgrade_from_file_path), object_pairs_hook=collections.OrderedDict)
+            upgrade_from['latest_version'][branch_nox] = version
+            repo.put_file(upgrade_from_file_path, json.dumps(upgrade_from, indent=4))
+            repo.commit('Add %s packages to packages_mapping' % version)
+        repo.push(new_branchname)
+        pr_text = self.github.create_pr(
+            target_repo='{}/{}'.format(upstream_name, repo_name),
+            target_branch='master',
+            source_user=self.username,
+            source_branch=new_branchname,
+            title='Add %s packages to packages_mapping' % branches,
+            text='')
+        self.slack.reply(pr_text, True)
+
+    def collect_packages(self, url):
+        """Given a release URL, returns a dict where keys are platform names, and values are {'url': 'http...'} """
+        release_data = requests.get(url).json()
+        # release_data['artifacts'] is a dictionary (called 'table'), each element is a list
+        # to flattern it:
+        # [item for table in release_data['artifacts'] for item in release_data['artifacts'][table]]
+        # to choose only items with defined platforms (they are called 'package' and might be undefined)
+        return {item['package']: {'url': item['URL']}
```

Comment:
A bonus here, although maybe unnecessary, would be to try to get 1 byte from the URL and check the status code to test that the URL is correct.

Reply:
I agree, but probably it should be done by some other script - for example, the one which checks SHAs.